tokio/src/sync/broadcast.rs

```diff
@@ -661,35 +664,61 @@ impl<T> Receiver<T> {
             // the slot lock.
             drop(slot);
 
-            let tail = self.shared.tail.lock().unwrap();
+            let mut tail = self.shared.tail.lock().unwrap();
 
             // Acquire slot lock again
             slot = self.shared.buffer[idx].read().unwrap();
 
-            // `tail.pos` points to the slot that the **next** send writes to. If
-            // the channel is closed, the previous slot is the oldest value.
-            let mut adjust = 0;
-            if tail.closed {
-                adjust = 1
-            }
-            let next = tail
-                .pos
-                .wrapping_sub(self.shared.buffer.len() as u64 + adjust);
+            // Make sure the position did not change. This could happen in the
+            // unlikely event that the buffer is wrapped between dropping the
+            // read lock and acquiring the tail lock.
+            if slot.pos != self.next {
+                let next_pos = slot.pos.wrapping_add(self.shared.buffer.len() as u64);
+
+                if next_pos == self.next {
+                    // Store the waker
+                    if let Some((waiter, waker)) = waiter {
+                        // Safety: called while locked.
+                        unsafe {
+                            // Only queue if not already queued
+                            waiter.with_mut(|ptr| {
+                                if !(*ptr).queued {
```

Comment:
This doesn't yet handle a `Waker` update if the `Future` gets polled again with a different `Waker`, right?

Reply:
Woops... yeah.

It would be really nice to somehow avoid having to re-implement all of these things.

Reply:
Fixed. thanks again.